src/device/execution.jl

```diff
@@ -1,110 +1,8 @@
-    end
-
-    # the number of elements to process needs to be passed to the kernel somehow, so there's
-    # no easy way to do this without passing additional arguments or changing the context.
-    # both are expensive, so require manual use of `launch_heuristic` for those kernels.
-    elements_per_thread = 1
-
-    if elements !== nothing
-        @assert elements > 0
-        heuristic = launch_heuristic(backend(target), kernel, args...;
-                                     elements, elements_per_thread)
-        config = launch_configuration(backend(target), heuristic;
-                                      elements, elements_per_thread)
-        gpu_call(backend(target), kernel, args, config.threads, config.blocks; name=name)
-    else
-        @assert threads > 0
-        @assert blocks > 0
-        gpu_call(backend(target), kernel, args, threads, blocks; name=name)
-    end
-end
-
-# how many threads and blocks `kernel` needs to be launched with, passing arguments `args`,
-# to fully saturate the GPU. `elements` indicates the number of elements that needs to be
-# processed, while `elements_per_threads` indicates the number of elements this kernel can
-# process (i.e. if it's a grid-stride kernel, or 1 if otherwise).
-#
-# this heuristic should be specialized for the back-end, ideally using an API for maximizing
-# the occupancy of the launch configuration (like CUDA's occupancy API).
-function launch_heuristic(backend::AbstractGPUBackend, kernel, args...;
-                          elements::Int, elements_per_thread::Int)
-    return (threads=256, blocks=32)
-end
-
-# determine how many threads and blocks to actually launch given upper limits.
-# returns a tuple of blocks, threads, and elements_per_thread (which is always 1
-# unless specified that the kernel can handle a number of elements per thread)
-function launch_configuration(backend::AbstractGPUBackend, heuristic;
-                              elements::Int, elements_per_thread::Int)
-    threads = clamp(elements, 1, heuristic.threads)
-    blocks = max(cld(elements, threads), 1)
-
-    if elements_per_thread > 1 && blocks > heuristic.blocks
-        # we want to launch more blocks than required, so prefer a grid-stride loop instead
-        ## try to stick to the number of blocks that the heuristic suggested
-        blocks = heuristic.blocks
-        nelem = cld(elements, blocks*threads)
-        ## only bump the number of blocks if we really need to
-        if nelem > elements_per_thread
-            nelem = elements_per_thread
-            blocks = cld(elements, nelem*threads)
-        end
-        (; threads, blocks, elements_per_thread=nelem)
-    else
-        (; threads, blocks, elements_per_thread=1)
-    end
-end
-
-gpu_call(backend::AbstractGPUBackend, kernel, args, threads::Int, blocks::Int; kwargs...) =
-    error("Not implemented") # COV_EXCL_LINE
+# TODO:
+# - Rename KA device to backend
+# - Who owns `AbstractGPUBackend`?
```

Comment:
Right now, `AbstractGPUBackend` is only used in the JLArrays lib, but can't we work towards deprecating that lib with KA's CPU capabilities?

If so, why do we need `AbstractGPUBackend`? Can't we just use KA's `Backend`?

Reply:
I think we need a JLArrays implementation using KA first.